---
title: "Preprocessing_DoubleDetection.Rmd"
author: "Mai La"
date: "2024-08-08"
output: html_document
---

# Background.

Doublet detection on Cellbender output on all samples.
Code run as described [here] and downstream analysis based on a tutorial from [`scCustomise`] (https://samuel-marsh.github.io/scCustomize/articles/Cell_Bender_Functions.html) and a tutorial from [`10X`](https://www.10xgenomics.com/cn/resources/analysis-guides/background-removal-guidance-for-single-cell-gene-expression-datasets-using-third-party-tools). 

Codes adapted from Fer's 'QC_Doublet_Detection.Rmd' code
# Prepare environment
```{r global-options, include=FALSE}
knitr::opts_chunk$set(echo = F, warning = F, error = F, message = F, results = "hide", eval = T)

bpp <- BiocParallel::MulticoreParam(workers = 12, 
                                    RNGseed = 1234,
                                    progressbar = TRUE)
options(future.globals.maxSize= 1048576000)
```

## Libraries.

```{r import_libraries, include = TRUE, echo=TRUE, results='hide', message=FALSE}
library("here")
library("Seurat")
library("dplyr")
library("RColorBrewer")
library("openxlsx")
library('schex')
library("EnhancedVolcano")
library("clusterProfiler")
library("org.Hs.eg.db")
library("ReactomePA")
library("cowplot")
library("openxlsx")
library("scCustomize")
library("scuttle")
library("dplyr")
library("SingleCellExperiment")
```

# Samples list and metadata.
```{r sample_metadata}
sample.metadata <- read.csv("~/projects/PDAC_non_malignant_cell_types/data/import.files/SG_sample_meta.csv", header = T, stringsAsFactors = F)

sts.samples <- dplyr::filter(sample.metadata, LTS != "Yes", Atlas == "Yes")

print(sts.samples)
```

# Load matrices
## Read sample files
```{r load_sce_object}
cellbender.out <- list.files(path = "/homevol/huemail/tmp/cellBender/samples/",
           pattern = "_cellbender_filtered.h5",
           full.names = T,
           recursive = T
            )

names(cellbender.out) <- cellbender.out %>% gsub(".*/|_cellbender_filtered.h5|-GEX", "", .)
```

## Discard low Q samples and LTS samples

```{r}
sts.samples.names <- sapply(sts.samples$Case, function(x)
  grep(x, names(cellbender.out), value = T)) %>% unlist() %>% as.character() %>% unique()

# Post-filtering low quality cells and discarding lowly expressed genes (see below), sample 2057-V1 only had 389 genes that were expressed in at least 25 cells and will be discarded. 
# Also, samples 2179, 2210, 2911 2221 and 2946.
# Patient #2212 has no malignant epithelial, also removed
# IPMN is from the Avner's cohort, also removed

low.q.samples <- c("2057-V1", "2179", "2210", "2911", "2221", "2946", 
                   "2212") 
intersect(low.q.samples, sts.samples.names)

sts.samples.names <- sts.samples.names[!sts.samples.names %in% low.q.samples]

cellbender.out <- cellbender.out[names(cellbender.out) %in% sts.samples.names]
# Please note that sample 2302 was run twice (different sections?) on runs STTM_110319, run on march 2019 (run 190322_A00130_0097) and STTM_14082019, run in August 2019 (190819_A00130_0114_AHMCLHDSX).
# As previously done, sample from run 190819_A00130_0114_AHMCLHDSX will be renamed as section 2.

names(cellbender.out)[21] <- "2302-section-2"

library(DropletUtils)
sts.sce <-
  read10xCounts(
    samples = cellbender.out,
    sample.names = names(cellbender.out),
    col.names = TRUE
  ) # should have 30 samples total

## Additional step to check that there no negative values or NAs
counts <- counts(sts.sce) %>% as.matrix()
any(counts < 0) # FALSE , confirm no negative value
any(is.na(counts)) # FALSE, confirm no NAs
```

# Discard samples where the majority of genes are not expressed.
```{r discard_samples_low_no_genes_expressed}
library(scater)
nexprs.list <- lapply(sts.sce$Sample %>% unique(),
                      function(x)
                        table(nexprs(sts.sce[, sts.sce$Sample == x], byrow = T) >= 25))

names(nexprs.list) <- sts.sce$Sample %>% unique()

nexprs.discard <- sapply(names(nexprs.list),
                         function(x)
                           nexprs.list[[x]][2] < 2500,
                         USE.NAMES = T)

# No samples to discard
print(names(which(nexprs.discard)) %>% gsub(".TRUE", "", .))
```

Without gene prefiltering none of the samples were discarded based on the criteria described above.

# Doublet detection
Doublet detection and conceptual framework adapted from [OSCA advanced chapter 8](http://bioconductor.org/books/3.13/OSCA.advanced/doublet-detection.html#further-comments-3).
Doublet detection could also be performed on a split object (see addendum).

## Preprocessing
### QC
Discard only very low quality cells. According to `scDblFinder` vignette discarding cells with < 200 UMIs is reasonable as an initial qc (more [here](https://bioconductor.org/packages/release/bioc/vignettes/scDblFinder/inst/doc/scDblFinder.html#usage)). 

```{r scater_QA/QC}

qc.dbl.df <- perCellQCMetrics(sts.sce)

qc.dbl.sum <- qc.dbl.df$sum < 200

table(qc.dbl.sum)
#qc.dbl.sum
# FALSE   TRUE 
#135982   4752 

#dim(sts.sce)
#36601 140734
```


### Discard low quality cells.
```{r discard_lowq_cells}
fltrd.sts.sce <- sts.sce[, !qc.dbl.sum]

table(nexprs(fltrd.sts.sce, byrow = T, BPPARAM = bpp) > 0)
#FALSE  TRUE 
# 2725 33876 
# Keep expressed genes only.

set.seed(1234)
non.zero.features <- nexprs(fltrd.sts.sce, byrow = T, BPPARAM = bpp) > 0
table(non.zero.features)
#non.zero.features
#FALSE  TRUE 
# 2725 33876

fltrd.sts.sce <- fltrd.sts.sce[non.zero.features, ]
```

### Discard samples with low no. of cells (~500 cells).
Discard samples with low no. of cells post initial QC.

```{r low_cell_samples_discard }
print(as.data.frame(colData(fltrd.sts.sce)) %>% dplyr::count(Sample) %>% dplyr::arrange(n))
# No samples with less than 500 cells.

saveRDS(fltrd.sts.sce,
  file = "~/projects/PDAC_non_malignant_cell_types/analysis/postCellBender_QCs/Rerun_Aug24/fltrd.sts.sce.RDS")
```


### Normalisation
```{r}
set.seed(1234)
dbl.clusters.samples <- scran::quickCluster(fltrd.sts.sce,
                                            BPPARAM = bpp,
                                            block=fltrd.sts.sce$Sample
                                            )

                             #BPPARAM = bpp, # no parallelization if block is not in used 
                             #block=cut(seq_len(ncol(fltrd.sts.sce)), 10), # additional line to help speed things up? 
                             #block=fltrd.sts.sce$Sample,
                             #method = "igraph")

# Notes on clustering within block: "We can break up the dataset by specifying block to cluster cells, usually within each batch or run. This generates clusters within each level of block, which is entirely adequate for applications like computeSumFactors where the aim of clustering is to separate dissimilar cells rather than group together all similar cells. Blocking reduces computational work considerably by allowing each level to be processed independently, without compromising performance provided that there are enough cells within each batch.
#Indeed, for applications like computeSumFactors, we can use block even in the absence of any known batch structure. Specifically, we can set it to an arbitrary factor such as block=cut(seq_len(ncol(x)), 10) to split the cells into ten batches of roughly equal size. This aims to improve speed, especially when combined with block.PARAM to parallelize processing of the independent levels" https://rdrr.io/bioc/scran/man/quickCluster.html

#More discussion here: https://support.bioconductor.org/p/9155953

saveRDS(dbl.clusters.samples,
  file = "~/projects/PDAC_non_malignant_cell_types/analysis/postCellBender_QCs/Rerun_Aug24/dbl.clusters.samples.RDS")

library(scuttle)
library(scran)
library(DropletUtils)
fltrd.sts.sce <-
  computeSumFactors(fltrd.sts.sce,
                    clusters = dbl.clusters.samples,
                    BPPARAM = bpp)

print(sizeFactors(fltrd.sts.sce) %>% head(n = 10))
# [1] 20.654276 12.916691 16.909326 13.964262  5.749571  6.198304  6.194301  7.970234  6.053636  5.061844

# To continue from here
fltrd.sts.sce <-
  logNormCounts(fltrd.sts.sce)

saveRDS(fltrd.sts.sce,
  file = "~/projects/PDAC_non_malignant_cell_types/analysis/postCellBender_QCs/Rerun_Aug24/fltrd.sts.sce.samples.RDS")

```

### Variance modelling
```{r variance_modelling} 
set.seed(1234)
dec.dbl <-
  modelGeneVarByPoisson(fltrd.sts.sce,
                        BPPARAM = bpp)
top.dec.dbl <-
  getTopHVGs(dec.dbl, n = 5000)

saveRDS(top.dec.dbl,
  file = "~/projects/PDAC_non_malignant_cell_types/analysis/postCellBender_QCs/Rerun_Aug24/top.dec.dbl.RDS")
saveRDS(dec.dbl,
  file = "~/projects/PDAC_non_malignant_cell_types/analysis/postCellBender_QCs/Rerun_Aug24/dec.dbl.RDS")
saveRDS(fltrd.sts.sce,
  file = "~/projects/PDAC_non_malignant_cell_types/analysis/postCellBender_QCs/Rerun_Aug24/fltrd.sts.sce.samples.RDS")
```

### Dimensionality reduction.
```{r pca}
set.seed(1234)
fltrd.sts.sce <-
  denoisePCA(fltrd.sts.sce,
             technical = dec.dbl,
             subset.row = top.dec.dbl) #this step take some time 1h30

reducedDimNames(fltrd.sts.sce)

sample.no <- length(unique(fltrd.sts.sce$Sample))
sample.colours  <- colorRampPalette(brewer.pal(11, "Spectral"))

plotReducedDim(fltrd.sts.sce, colour_by = "Sample", dimred = "PCA") + scale_color_manual(values = sample.colours(sample.no))

set.seed(1234)
fltrd.sts.sce <- runUMAP(
  fltrd.sts.sce,
  dimred = "PCA"#,
  # external_neighbors = TRUE,
  # BNPARAM = BiocNeighbors::AnnoyParam(),
  # BPPARAM = bpp,
  # n_threads = 16
  # n_threads = BiocParallel::bpnworkers(bpp)
)
 
reducedDimNames(fltrd.sts.sce)

plotReducedDim(fltrd.sts.sce,
               colour_by = "Sample",
               dimred = "UMAP"#,
               # text_by = "Sample",
               # text_col = "black"
               ) + scale_color_manual(values = sample.colours(sample.no))

plotReducedDim(fltrd.sts.sce,
               colour_by = "MUC2",
               swap_rownames = "Symbol",
               dimred = "UMAP",
               text_by = "Sample",
               # text_col = "black"
               )
# Both 2302 and IPMN highly express MUC2
```

### Clustering

```{r clustering}
require('scDblFinder')
set.seed(1234)

fltrd.sts.sce$fastClusters <-
  fastcluster(fltrd.sts.sce, rdname = "PCA") %>% as.factor()

colLabels(fltrd.sts.sce) <- fltrd.sts.sce$fastClusters

# Read object generated by `dbl_clustering.R` script.

table(colLabels(fltrd.sts.sce))

fltrd.sts.sce$Sample <- factor(fltrd.sts.sce$Sample, levels = fltrd.sts.sce$Sample %>% unique())

plotUMAP(fltrd.sts.sce, colour_by = "label") +
plotUMAP(fltrd.sts.sce, colour_by = "Sample")  + scale_color_manual(values = sample.colours(sample.no))
```

## scDblFinder
### Run scDblFinder
```{r scdDblFinder}
require(scDblFinder)
# Seed set on MultiCoreParam
# set.seed(1234)

fltrd.sts.sce <-
  scDblFinder(
    fltrd.sts.sce,
    samples = "Sample",
    clusters = "fastClusters",
    BPPARAM = bpp
  )

table(fltrd.sts.sce$scDblFinder.class)
#singlet doublet 
# 129134    6848 

plotUMAP(fltrd.sts.sce, colour_by="scDblFinder.class") +
  plotUMAP(fltrd.sts.sce, colour_by="Sample") #+ scale_color_manual(values = sample.colours(sample.no))

saveRDS(fltrd.sts.sce,
  file = "~/projects/PDAC_non_malignant_cell_types/analysis/postCellBender_QCs/Rerun_Aug24/fltrd.sts.sce.RDS")
```

### Doublet detection metrics.
```{r dbl_metrics}
dbl.class.table.list <-
  lapply(fltrd.sts.sce$Sample %>% unique(), function(x)
    fltrd.sts.sce[, fltrd.sts.sce$Sample == x, ]$scDblFinder.class %>% tibble::as_tibble() %>%
      group_by(value) %>% tally()
    )

names(dbl.class.table.list) <- fltrd.sts.sce$Sample %>% unique()

library(purrr)
dbl.tibble <- map_dfr(dbl.class.table.list, bind_rows, .id = "Sample")
dbl.tibble <- dbl.tibble %>% group_by(Sample) %>% mutate( percentage = 100 * n /sum(n)) %>% ungroup()

dbl.tibble %>%
  filter(value == "doublet") %>%
  arrange(desc(percentage)) %>% 
  print(n=30)

print(as.data.frame(colData(fltrd.sts.sce)) %>%
        dplyr::count(Sample) %>%
        dplyr::arrange(desc(n)))
```

# Discard doublets
```{r}
dbl.fltrd.sts.sce <-
  fltrd.sts.sce[, fltrd.sts.sce$scDblFinder.class == "singlet"]

print(as.data.frame(colData(dbl.fltrd.sts.sce)) %>%
        dplyr::count(scDblFinder.class) %>%
        dplyr::arrange(desc(n)))

saveRDS(dbl.fltrd.sts.sce,
  file = "~/projects/PDAC_non_malignant_cell_types/analysis/postCellBender_QCs/Rerun_Aug24/dbl.fltrd.sts.sce.RDS")
# this object has 129134 cells
# table(fltrd.sts.sce$scDblFinder.class)
# singlet doublet 
# 129134    6848 
# table(dbl.fltrd.sts.sce$scDblFinder.class)
# singlet doublet 
# 129134       0
```
